File: tests/equivalence/tableGoldens.R
#!/usr/bin/env Rscript
# Golden-table matrix generator (R side) for the Phase-2 Python equivalence work.
#
# Each case builds a jaspTable through the RCPP_MODULE ingest path (the same
# code real JASP modules use) and dumps the results JSON + toHtml. The Python
# twin (tableGoldens.py) builds the *same logical table* through the pybind
# adapter; the two outputs must be byte-identical (see runTableGoldens.sh).
#
# Usage:
#   JASP_EQUIV_RLIB=/path/to/lib Rscript tableGoldens.R [outdir]

if (Sys.getenv("JASP_EQUIV_RLIB") != "")
	.libPaths(c(Sys.getenv("JASP_EQUIV_RLIB"), .libPaths()))

suppressPackageStartupMessages(library(jaspBase))

ns <- asNamespace("jaspBase")
get0 <- function(n) get(n, envir=ns)

outdir <- {
	args <- commandArgs(trailingOnly=TRUE)
	if (length(args) >= 1) args[[1]] else file.path(getwd(), "tests", "equivalence", "tableGoldens")
}
dir.create(outdir, showWarnings=FALSE, recursive=TRUE)

get0("setDeveloperMode")(FALSE)

# Each case fills `t` (a fresh jaspTable). The harness wraps it in a jaspResults
# so we exercise the same dataEntry/meta/JSON machinery as production.
runCase <- function(name, buildTable)
{
	get0("destroyAllAllocatedObjects")()
	results <- get0("create_cpp_jaspResults")("goldens", NULL)
	get0("setResponseData")(7L, 0L)
	results$setOptions('{"case":"placeholder"}')

	t <- get0("create_cpp_jaspTable")(name)
	buildTable(t)
	results[["table"]] <- t

	writeLines(results$getResults(), file.path(outdir, paste0(name, "_results.json")))
	writeLines(t$toHtml(),           file.path(outdir, paste0(name, "_toHtml.txt")))
	cat("  R case ok:", name, "\n")
}

# --- cases -------------------------------------------------------------------

runCase("01_numeric_string", function(t) {
	t$setData(list(x=c(1.5, NA, Inf), s=c("a<b", "c&d", NA)))
	t$status <- "complete"
})

runCase("02_int_lgl", function(t) {
	t$setData(list(i=c(1L, 2L, NA), b=c(TRUE, NA, FALSE)))
	t$status <- "complete"
})

runCase("03_nan", function(t) {
	t$setData(list(v=c(1.0, NaN, NA)))
	t$status <- "complete"
})

runCase("04_row_names", function(t) {
	t$setData(data.frame(x=c(1,2,3), row.names=c("r1","r2","r3")))
	t$status <- "complete"
})

runCase("05_col_metadata", function(t) {
	t$setData(list(a=c(1,2), b=c("x","y")))
	t$setColTitles(list("Alpha", "Beta"))
	t$setColTypes(list("number", "string"))
	t$setColFormats(list("sf:4", ""))
	t$status <- "complete"
})

runCase("06_footnotes", function(t) {
	t$setData(list(x=c(1,2), y=c(3,4)))
	t$addFootnoteHelper("table note", NULL, NULL, NULL)
	t$addFootnoteHelper("col note", "*", "x", NULL)
	t$addFootnoteHelper("cell note", "#", "y", "2")
	t$status <- "complete"
})

runCase("07_add_column_info", function(t) {
	t$addColumnInfoHelper("m", "Mean", "number", "sf:4", TRUE, "Stats")
	t$addColumnInfoHelper("p", "p", "pvalue", "dp:3", NULL, "Stats")
	t$setData(list(m=c(1.234), p=c(0.041)))
	t$status <- "complete"
})

runCase("08_mixed_column", function(t) {
	mx <- createMixedColumn(list(1.5, "txt", 2L), c("number", "string", "integer"), list("sf:4", NULL, ""))
	t[["mx"]] <- mx
	t$status <- "complete"
})

runCase("09_add_rows", function(t) {
	t$setData(data.frame(x=numeric(0), y=numeric(0)))
	t$addRows(list(c(1, 10), c(2, 20), c(3, 30)))
	t$status <- "complete"
})

runCase("10_add_row_named", function(t) {
	t$setData(data.frame(x=numeric(0), y=numeric(0)))
	t$addRow(c(5, 6), "onlyrow")
	t$status <- "complete"
})

runCase("11_add_columns", function(t) {
	t$setData(list(a=c(1,2)))
	t$addColumns(list(b=c(3,4), c=c("u","v")))
	t$status <- "complete"
})

runCase("12_transpose", function(t) {
	t$setData(list(x=c(1,2), y=c(3,4)))
	t$setRowNames(list("r1","r2"))
	t$transpose <- TRUE
	t$status <- "complete"
})

runCase("13_expected_size", function(t) {
	t$setExpectedSize(3, 2)
	t$setData(list(x=c(1,2)))
	t$status <- "running"
})

runCase("14_matrix_input", function(t) {
	# matrix() fills column-major: same logical layout as the Python twin
	# (np.array rows [[1,4],[2,5],[3,6]] -> columns c1=[1,2,3], c2=[4,5,6]).
	m <- matrix(c(1, 2, 3, 4, 5, 6), nrow=3, ncol=2)
	colnames(m) <- c("c1", "c2")
	t$setData(m)
	t$status <- "complete"
})

runCase("15_vector_row", function(t) {
	t$setData(c(1.5, 2.5, 3.5))
	t$status <- "complete"
})

runCase("16_list_of_columns", function(t) {
	t$setData(list(a=c(1,2,3), b=c("x","y","z")))
	t$status <- "complete"
})

runCase("17_escape_and_citation", function(t) {
	t$setData(list(s=c("Tom & Jerry", "<script>", "1 < 2")))
	t$addCitation("A citation with <tags> & symbols")
	t$status <- "complete"
})

runCase("18_empty_table", function(t) {
	t$status <- "complete"
})

runCase("19_dataframe_rownames", function(t) {
	# data.frame carries implicit row names "1".."n"; the Python twin passes
	# them explicitly (documented §3.2 difference: no implicit index names).
	t$setData(data.frame(x=c(1,2,3), s=c("a","b","c"), stringsAsFactors=FALSE))
	t$status <- "complete"
})

runCase("20_factor_categorical", function(t) {
	t$setData(data.frame(f=factor(c("a","b",NA,"a")), v=c(1,2,3,4)))
	t$status <- "complete"
})

cat("R goldens written to", outdir, "\n")

File: tests/equivalence/tableGoldens.py
#!/usr/bin/env python3
"""Golden-table matrix generator (Python side) for the Phase-2 equivalence work.

Twin of tableGoldens.R: each case builds the *same logical table* through the
pybind adapter's Python-native ingest (dict / list / ndarray) that the R side
builds through data.frame / list / matrix. Outputs must be byte-identical
(see runTableGoldens.sh).

Requires the built `_jaspresults` extension; point JASP_PY_MODULE_DIR at its
directory (defaults to python/build).
"""
import os
import sys

mod_dir = os.environ.get(
	"JASP_PY_MODULE_DIR",
	os.path.join(os.path.dirname(__file__), "..", "..", "python", "build"))
sys.path.insert(0, os.path.abspath(mod_dir))

import _jaspresults as M  # noqa: E402

outdir = sys.argv[1] if len(sys.argv) >= 1 + 1 else os.path.join(
	os.path.dirname(__file__), "tableGoldens")
os.makedirs(outdir, exist_ok=True)

M.setDeveloperMode(False)


def run_case(name, build_table):
	M.destroyAllAllocatedObjects()
	results = M.jaspResults("goldens")
	M.setResponseData(7, 0)
	results.setOptions('{"case":"placeholder"}')

	t = M.jaspTable(name)
	build_table(t)
	results.insert("table", t)

	with open(os.path.join(outdir, name + "_results.json"), "w") as f:
		f.write(results.getResults() + "\n")
	with open(os.path.join(outdir, name + "_toHtml.txt"), "w") as f:
		f.write(t.toHtml() + "\n")


def case_01(t):
	t.setData({"x": [1.5, None, float("inf")], "s": ["a<b", "c&d", None]})
	t.status = "complete"


def case_02(t):
	t.setData({"i": [1, 2, None], "b": [True, None, False]})
	t.status = "complete"


def case_03(t):
	t.setData({"v": [1.0, M.NaNString, None]})
	t.status = "complete"


def case_04(t):
	# R c(1,2,3) is double; use floats so the cell type matches ("number").
	t.setData({"x": [1.0, 2.0, 3.0]}, row_names=["r1", "r2", "r3"])
	t.status = "complete"


def case_05(t):
	t.setData({"a": [1.0, 2.0], "b": ["x", "y"]})
	t.setColTitles(["Alpha", "Beta"])
	t.setColTypes(["number", "string"])
	t.setColFormats(["sf:4", ""])
	t.status = "complete"


def case_06(t):
	t.setData({"x": [1.0, 2.0], "y": [3.0, 4.0]})
	t.addFootnote("table note")
	t.addFootnote("col note", symbol="*", col_names=["x"])
	t.addFootnote("cell note", symbol="#", col_names=["y"], row_names=["2"])
	t.status = "complete"


def case_07(t):
	t.addColumnInfo("m", "Mean", "number", "sf:4", True, "Stats")
	t.addColumnInfo("p", "p", "pvalue", "dp:3", None, "Stats")
	t.setData({"m": [1.234], "p": [0.041]})
	t.status = "complete"


def case_08(t):
	t.setColumn("mx", [
		{"value": 1.5, "type": "number", "format": "sf:4"},
		{"value": "txt", "type": "string", "format": None},
		{"value": 2, "type": "integer", "format": ""},
	])
	t.status = "complete"


def case_09(t):
	t.setData({"x": [], "y": []})
	t.addRows([[1.0, 10.0], [2.0, 20.0], [3.0, 30.0]])
	t.status = "complete"


def case_10(t):
	t.setData({"x": [], "y": []})
	t.addRow([5.0, 6.0], row_name="onlyrow")
	t.status = "complete"


def case_11(t):
	t.setData({"a": [1.0, 2.0]})
	t.addColumns({"b": [3.0, 4.0], "c": ["u", "v"]})
	t.status = "complete"


def case_12(t):
	t.setData({"x": [1.0, 2.0], "y": [3.0, 4.0]})
	t.setRowNames(["r1", "r2"])
	t.transpose = True
	t.status = "complete"


def case_13(t):
	t.setExpectedSize(3, 2)
	t.setData({"x": [1.0, 2.0]})
	t.status = "running"


def case_14(t):
	# 2-D ndarray -> columns (like an R matrix)
	np = __import__("numpy")
	arr = np.array([[1, 4], [2, 5], [3, 6]], dtype=float)
	t.setData(arr, col_names=["c1", "c2"])
	t.status = "complete"


def case_15(t):
	t.setData([1.5, 2.5, 3.5])
	t.status = "complete"


def case_16(t):
	t.setData({"a": [1.0, 2.0, 3.0], "b": ["x", "y", "z"]})
	t.status = "complete"


def case_17(t):
	t.setData({"s": ["Tom & Jerry", "<script>", "1 < 2"]})
	t.addCitation("A citation with <tags> & symbols")
	t.status = "complete"


def case_18(t):
	t.status = "complete"


def case_19(t):
	# R data.frame carries implicit row names "1".."n"; Python passes them
	# explicitly (documented §3.2 difference: no implicit index names).
	t.setData({"x": [1.0, 2.0, 3.0], "s": ["a", "b", "c"]}, row_names=["1", "2", "3"])
	t.status = "complete"


def case_20(t):
	# R factor -> character column; pandas Categorical -> category values.
	# R side also has a numeric column v=c(1,2,3,4) (doubles).
	pd = __import__("pandas")
	df = pd.DataFrame({
		"f": pd.Categorical(["a", "b", None, "a"]),
		"v": [1.0, 2.0, 3.0, 4.0],
	})
	# R data.frames carry implicit row names "1".."n"; pass them explicitly.
	t.setData(df, row_names=["1", "2", "3", "4"])
	t.status = "complete"


CASES = {
	"01_numeric_string": case_01,
	"02_int_lgl": case_02,
	"03_nan": case_03,
	"04_row_names": case_04,
	"05_col_metadata": case_05,
	"06_footnotes": case_06,
	"07_add_column_info": case_07,
	"08_mixed_column": case_08,
	"09_add_rows": case_09,
	"10_add_row_named": case_10,
	"11_add_columns": case_11,
	"12_transpose": case_12,
	"13_expected_size": case_13,
	"14_matrix_input": case_14,
	"15_vector_row": case_15,
	"16_list_of_columns": case_16,
	"17_escape_and_citation": case_17,
	"18_empty_table": case_18,
	"19_dataframe_rownames": case_19,
	"20_factor_categorical": case_20,
}

for name, fn in CASES.items():
	run_case(name, fn)

print("Python goldens written to", outdir)
